constantine/constantine-jni/src/main/c/ethereum_evm_precompiles.c

```diff
@@ -0,0 +1,69 @@
+#include <jni.h>
+#include "org_hyperledger_besu_nativelib_constantine_LibConstantineEIP196.h"
+#include <constantine.h>
+#include <stdio.h>
+
+void printByteArray(const char* label, const byte* array, size_t len) {
+    printf("%s: [", label);
+    for (size_t i = 0; i < len; i++) {
+        printf("%02x", array[i]);
+        if (i < len - 1) {
+            printf(", ");
+        }
+    }
+    printf("]\n");
+}
+
+JNIEXPORT jint JNICALL Java_Constantine_ctt_1eth_1evm_1bn254_1g1add(JNIEnv *env, jobject obj, jbyteArray jr, jint r_len, jbyteArray jinputs, jint inputs_len) {
+    jbyte *r = (*env)->GetByteArrayElements(env, jr, NULL);
+    jbyte *inputs = (*env)->GetByteArrayElements(env, jinputs, NULL);
+
+    ctt_evm_status status = ctt_eth_evm_bn254_g1add((byte *)r, (ptrdiff_t)r_len, (const byte *)inputs, (ptrdiff_t)inputs_len);
+
+    if (status != cttEVM_Success) {
+        printf("ctt_eth_evm_bn254_g1add failed with status: %d\n", status);
+    } else {
+        printByteArray("Result", (const byte *)r, r_len);
+    }
+
+    (*env)->ReleaseByteArrayElements(env, jr, r, 0);
+    (*env)->ReleaseByteArrayElements(env, jinputs, inputs, 0);
+
+    return (jint)status;
+}
+
+JNIEXPORT jint JNICALL Java_Constantine_ctt_1eth_1evm_1bn254_1g1mul(JNIEnv *env, jobject obj, jbyteArray jr, jint r_len, jbyteArray jinputs, jint inputs_len) {
+    jbyte *r = (*env)->GetByteArrayElements(env, jr, NULL);
+    jbyte *inputs = (*env)->GetByteArrayElements(env, jinputs, NULL);
+
+    ctt_evm_status status = ctt_eth_evm_bn254_g1mul((byte *)r, (ptrdiff_t)r_len, (const byte *)inputs, (ptrdiff_t)inputs_len);
```

Comment:
very tidy that constantine already has precompile friendly endpoints 👍

Reply:
Yes, and SHA256 and MODEXP. The plan is to implement all cryptographic precompiles.

For the record, Constantine was actually started with the idea of having a clean from scratch BN254 backend for Nim but it was too hard/time-consuming at the time so a 1-1 port of https://github.com/zcash-hackworks/bn was made instead (https://github.com/status-im/nim-bncurve).
However performance is bad as even a simple add with carry instruction is transformed into 7 instructions: https://github.com/zcash-hackworks/bn/blob/master/src/arith.rs#L396-L405

The issue of zcash's `bn` were already mentioned in 2008 in EIP-1108 https://eips.ethereum.org/EIPS/eip-1108 that reduced gas cost price. And moving to Cloudflare bn256 improved speed by 10x:

> and corroborated in [a later note](https://github.com/ethereum/go-ethereum/pull/16301#issuecomment-372687543), the computational cost of ECADD, ECMUL, and pairing checks (excepting the constant) has dropped roughly an order of magnitude across the board.

However that Cloudflare library was about 2x slower than state-of-the-art in 2021 https://hackmd.io/@gnark/eccbench
https://user-images.githubusercontent.com/22738317/162632923-b057c0a2-6929-419e-b0de-4fd187a5e508.png
and Constantine there had a parameter passing bug which made it 2x slower than necessary on pairings (https://github.com/nim-lang/Nim/issues/16897)

See full context in https://ethresear.ch/t/releasing-constantine-v0-1-0-a-modular-cryptography-stack-for-ethereum/19990